src/phoenix/server/api/subscriptions.py

```diff
@@ -219,89 +242,75 @@ async def chat_completion_over_dataset(
+                or _default_playground_experiment_description(dataset_name=dataset.name),
                 repetitions=1,
                 metadata_=input.experiment_metadata
                 or _default_playground_experiment_metadata(
-                    dataset_name=dataset_name,
+                    dataset_name=dataset.name,
                     dataset_id=input.dataset_id,
-                    version_id=resolved_version_node_id,
+                    version_id=GlobalID(DatasetVersion.__name__, str(resolved_version_id)),
                 ),
                 project_name=PLAYGROUND_PROJECT_NAME,
             )
             session.add(experiment)
             await session.flush()
-            runs = [
-                models.ExperimentRun(
-                    experiment_id=experiment.id,
-                    dataset_example_id=from_global_id_with_expected_type(
-                        example_id, DatasetExample.__name__
-                    ),
-                    trace_id=span.trace_id,
-                    output=models.ExperimentRunOutput(
-                        task_output=_get_playground_experiment_task_output(span)
-                    ),
-                    repetition_number=1,
-                    start_time=span.start_time,
-                    end_time=span.end_time,
-                    error=error_message
-                    if (error_message := span.error_message) is not None
-                    else None,
-                    prompt_token_count=get_attribute_value(span.attributes, LLM_TOKEN_COUNT_PROMPT),
-                    completion_token_count=get_attribute_value(
-                        span.attributes, LLM_TOKEN_COUNT_COMPLETION
-                    ),
-                )
-                for example_id, span in spans.items()
-            ]
-            session.add_all(runs)
-            await session.flush()
-        for example_id in spans:
-            yield FinishedChatCompletion(
-                span=to_gql_span(db_spans[example_id]),
-                dataset_example_id=example_id,
+        yield ChatCompletionSubscriptionExperiment(experiment=to_gql_experiment(experiment))
+
+        results_queue: Queue[ChatCompletionResult] = Queue()
+        chat_completion_iterators = [
+            _chat_completion_payloads(
+                input=input,
+                llm_client_class=llm_client_class,
+                revision=revision,
+                results_queue=results_queue,
+                experiment_id=experiment.id,
+                project_id=playground_project_id,
             )
-        yield ChatCompletionOverDatasetSubscriptionResult(experiment=to_gql_experiment(experiment))
+            for revision in revisions
+        ]
+        tasks: dict[
+            AsyncIterator[ChatCompletionSubscriptionPayload],
+            Task[ChatCompletionSubscriptionPayload],
+        ] = {iterator: _as_task(iterator) for iterator in chat_completion_iterators}
+        batch_size = 10
+        while tasks:
```

Comment:
This logic is fairly intricate, would some variable renames, inline comments or a brief overview of the technique suffice in making the approach more clear on inspection?

Reply:
sync offline, addressed in commit below